internal/xds/rbac/converter.go

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package rbac provides service-level and method-level access control for a
+// service. See
+// https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#role-based-access-control-rbac
+// for documentation.
+package rbac
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v1xdsudpatypepb "github.com/cncf/xds/go/udpa/type/v1"
+	v3xdsxdstypepb "github.com/cncf/xds/go/xds/type/v3"
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	"google.golang.org/grpc/authz/audit"
+	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func buildLogger(loggerConfig *v3rbacpb.RBAC_AuditLoggingOptions_AuditLoggerConfig) (audit.Logger, error) {
```

Comment:
Can we add unit tests for this func and the one below?

Reply:
I'm not against testing these individual functions, but I'm not sure it fits with the repos existing style. Typically what I've seen so far is only testing the exported functions.
All of these helpers are effectively tested by `TestNewChainEngine` and `TestChainEngine`

Thoughts?

Reply:
The `converter.go` of xdslbregistry has its own [test](https://github.com/easwars/grpc-go/blob/master/xds/internal/xdsclient/xdslbregistry/tests/converter_test.go). But I agree since it's a standalone pkg so it's not entirely the same case here.

That said, I don't think all the possible errors in `getCustomConfig` are covered by those you mentioned. If you can add test cases there, it'll be fine too. Or please let me know which tests are covering those errors right now if I missed them.

Reply:
The existing tests make it extremely difficult to see what is tested where, I think adding excessive unit tests on this might be useful if just for readability

Reply:
SGTM. I think the `buildLogger` is likely fine. It's really for the `getCustomConfig`. I should've commented on a different line. But thanks!

Reply:
Also to add, the existing test only checks whether there is an error but not what error. I suppose it's easier that way there. That's when the unit tests for internal APIs where we can check the error string become useful.

Reply:
I was gonna say it's fine not to cover it. But looks like you did it already, which I think is great. Thanks!

Reply:
To clarify that one specific case is not covered
Also, agree on the existing tests only checking whether there is an error, not that it's the correct one - since changing that requires changing the existing tests, I'd rather fix that in a separate PR.